### PR TITLE
Intake

### PR DESCRIPTION
**Pull request by sunaynas**
ready for code review

### CODE REVIEW

File: src/main/cpp/Intake.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        case STOWED:
+            m_wristMotor.SetVoltage(units::volt_t(std::clamp(m_stowedPIDcontroller.Calculate(IntakeConstants::STOWED_POS - m_curPos), -IntakeConstants::WRIST_MAX_VOLTS, IntakeConstants::WRIST_MAX_VOLTS)));
+            m_rollerMotor.SetVoltage(units::volt_t(0));
+            break;
+        case STOWING:
+        case DEPLOYING:
+            UpdateTargetPose(); // bc still using motion profile 
+            m_wristMotor.SetVoltage(units::volt_t(std::clamp(FFPIDCalculate(), -IntakeConstants::WRIST_MAX_VOLTS, IntakeConstants::WRIST_MAX_VOLTS)));
+            if (AtSetpoint()){
+                if (m_state == STOWING) m_state = STOWED;
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp:65]: Reset PID aswell here (to avoid I term being crzy)

File: src/main/cpp/Intake.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            m_rollerMotor.SetVoltage(units::volt_t(0));
+            break;
+        case STOWED:
+            m_wristMotor.SetVoltage(units::volt_t(std::clamp(m_stowedPIDcontroller.Calculate(IntakeConstants::STOWED_POS - m_curPos), -IntakeConstants::WRIST_MAX_VOLTS, IntakeConstants::WRIST_MAX_VOLTS)));
+            m_rollerMotor.SetVoltage(units::volt_t(0));
+            break;
+        case STOWING:
+        case DEPLOYING:
+            UpdateTargetPose(); // bc still using motion profile 
+            m_wristMotor.SetVoltage(units::volt_t(std::clamp(FFPIDCalculate(), -IntakeConstants::WRIST_MAX_VOLTS, IntakeConstants::WRIST_MAX_VOLTS)));
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp]: Might want to start up the roller here because it takes time to accelerate

Comment (linc-mvla) [reply src/main/cpp/Intake.cpp]: idk if you want to set voltage to roller for stowing

File: src/main/cpp/Intake.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            m_rollerMotor.SetVoltage(units::volt_t(std::clamp(m_rollerVolts, -IntakeConstants::ROLLER_MAX_VOLTS,IntakeConstants::ROLLER_MAX_VOLTS)));
+            break;
+    }
+}
+
+double Intake::FFPIDCalculate(){
+    double posErr = m_targetPos - m_curPos, 
+    velErr = m_targetVel - m_curVel;
+    m_totalErr += posErr;
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp]: Multiply by 0.02 because integrate

File: src/main/cpp/Intake.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    double pid = m_kp*posErr + m_kd*velErr + m_ki*m_totalErr;
+    double ff = m_g*cos(m_targetPos) + m_s*((m_targetVel > 0) ? 1.0 : -1.0 ) + m_v*m_targetVel + m_a*m_targetAcc;
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp]: simpliftye

File: src/main/cpp/Intake.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+      m_setPt = setpt;
+      m_targetPos = m_curPos;
+      m_targetVel = 0.0;
+      m_targetAcc = IntakeConstants::WRIST_MAX_ACC;
+      if (m_setPt < m_curPos) m_targetAcc *= -1;
+      ResetPID();
+      CalcVelTurnPos();
+}
+
+bool Intake::AtSetpoint(){
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp:145]: simplufy

File: src/main/cpp/Intake.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+// Converts steps to radians for the hex bore encoder
+double Intake::StepsToRad(double steps){
+  return steps * (2.0 * M_PI / 2048.0); // falcon is 2048 ticks per revolution
+}
+
+void Intake::ResetPID(){
+    m_totalErr = 0;
+}
+
+void Intake::CalcVelTurnPos(){
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp]: very weird name

Comment (linc-mvla) [reply src/main/cpp/Intake.cpp]: call it CalcProfileMaxVel or smthg

Comment (linc-mvla) [reply src/main/cpp/Intake.cpp]: prolly not that, DeaccleratePos is better

File: src/main/cpp/Intake.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+void Intake::UpdateTargetPose(){
+    if (dbg){
+        frc::SmartDashboard::PutNumber("targ pos", m_targetPos);
+        frc::SmartDashboard::PutNumber("targ vel", m_targetVel);
+        // frc::SmartDashboard::PutNumber("targ acc", m_targetAcc);
+    }
+    double newP = m_targetPos, newV = m_targetVel, newA = m_targetAcc;
+    
+    newP += m_targetVel;
```
Comment (linc-mvla) [inline src/main/cpp/Intake.cpp]: multiply by 0.02 to integrate

**Review by linc-mvla — CHANGES_REQUESTED** [review]
YEAH

File: src/main/cpp/Intake.cpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        frc::SmartDashboard::PutNumber("s", m_s * m_targetPos); 
+        frc::SmartDashboard::PutNumber("v", m_v * m_targetVel); 
+        frc::SmartDashboard::PutNumber("a", m_a * m_targetAcc); 
+    }
+    return pid+ff;
+}
+
+void Intake::ChangeWristPos(double newPos){
+    SetSetpoint(newPos);
+    m_state = DEPLOYING;
```
Comment (jonathanhliu21) [inline src/main/cpp/Intake.cpp:105]: Calling this shouldn't start deploying, should only start deploying when deploy method is called

**Review by jonathanhliu21 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**